File: ts/packages/luzid-sdk/src/api-types/cluster.ts
import { Cluster as GrpcCluster, clusterToJSON } from '@luzid/grpc'
import { assert } from '../core/assert'

const SOLANA_ENDPOINTS = [
  'http://127.0.0.1:8899',
  'https://api.testnet.solana.com',
  'https://api.mainnet-beta.solana.com',
  'https://api.devnet.solana.com',
] as const

export type SolanaEndpoint = (typeof SOLANA_ENDPOINTS)[number]

/**
 * A Solana cluster to be used with the Luzid SDK.
 * @see {@link Cluster}
 */
export type LuzidCluster = {
  apiUrl: SolanaEndpoint
  toJSON(): string
}

/**
 * Wrapper around the Cluster enum which provides a URL for each cluster including
 * the development cluster that Luzid is running.
 * This is similar to the @solana/web3.js Cluster except that the web3 version does
 * not include the development cluster.
 */
export class Cluster implements LuzidCluster {
  private constructor(
    /**
     * The underlying grpc cluster used by Luzid only visible to the Luzid SDK
     * itself.
     * @private
     */
    readonly grpcCluster: GrpcCluster
  ) {}

  /**
   * Local/development cluster, use this for transactions and queries against
   * the  locally running validator running inside the Luzid App.
   */
  static Development: LuzidCluster = new Cluster(GrpcCluster.Development)
  /** Solana testnet cluster */
  static Testnet: LuzidCluster = new Cluster(GrpcCluster.Testnet)
  /** Solana mainnet cluster */
  static MainnetBeta: LuzidCluster = new Cluster(GrpcCluster.MainnetBeta)
  /** Solana devnet cluster */
  static Devnet: LuzidCluster = new Cluster(GrpcCluster.Devnet)

  /**
   * Gets the URL for the specific Solana cluster.
   */
  get apiUrl(): SolanaEndpoint {
    switch (this.grpcCluster) {
      case GrpcCluster.Development:
        return 'http://127.0.0.1:8899'
      case GrpcCluster.Testnet:
        return 'https://api.testnet.solana.com'
      case GrpcCluster.MainnetBeta:
        return 'https://api.mainnet-beta.solana.com'
      case GrpcCluster.Devnet:
        return 'https://api.devnet.solana.com'
      default:
        throw new Error(`Unknown cluster: ${this.grpcCluster}`)
    }
  }

  toJSON() {
    return clusterToJSON(this.grpcCluster)
  }
}

/** @private */
export function isCluster(cluster: LuzidCluster): cluster is Cluster {
  return (cluster as Cluster).grpcCluster != null
}

/** @private */
export function asCluster(cluster: LuzidCluster): Cluster {
  assert(isCluster(cluster), 'Expected a Cluster')
  return cluster
}

export function intoGrpcCluster(cluster: LuzidCluster): GrpcCluster {
  return asCluster(cluster).grpcCluster
}

File: dart/luzid-sdk/lib/src/api-types/cluster.dart
import 'package:luzid_grpc/luzid_grpc.dart' as grpc;

final solanaEndpoints = [
  'http://127.0.0.1:8899',
  'https://api.testnet.solana.com',
  'https://api.mainnet-beta.solana.com',
  'https://api.devnet.solana.com'
];

class LuzidCluster {
  /// The underlying grpc cluster used by Luzid only visible to the Luzid SDK
  final grpc.Cluster _grpcCluster;

  LuzidCluster._(this._grpcCluster);

  /// Local/development cluster
  static LuzidCluster development = LuzidCluster._(grpc.Cluster.Development);

  /// Solana testnet cluster
  static LuzidCluster testnet = LuzidCluster._(grpc.Cluster.Testnet);

  /// Solana mainnet cluster
  static LuzidCluster mainnetBeta = LuzidCluster._(grpc.Cluster.MainnetBeta);

  /// Solana devnet cluster
  static LuzidCluster devnet = LuzidCluster._(grpc.Cluster.Devnet);

  /// Gets the URL for the specific Solana cluster.
  String get apiUrl {
    switch (_grpcCluster) {
      case grpc.Cluster.Development:
        return solanaEndpoints[0];
      case grpc.Cluster.Testnet:
        return solanaEndpoints[1];
      case grpc.Cluster.MainnetBeta:
        return solanaEndpoints[2];
      case grpc.Cluster.Devnet:
        return solanaEndpoints[3];
      default:
        throw Exception('Unknown cluster: $_grpcCluster');
    }
  }

  String get label {
    switch (_grpcCluster) {
      case grpc.Cluster.Development:
        return 'Development';
      case grpc.Cluster.Testnet:
        return 'Testnet';
      case grpc.Cluster.MainnetBeta:
        return 'MainnetBeta';
      case grpc.Cluster.Devnet:
        return 'Devnet';
      default:
        throw Exception('Unknown cluster: $_grpcCluster');
    }
  }

  grpc.Cluster intoGrpcCluster() {
    return _grpcCluster;
  }

  factory LuzidCluster.fromGrpc(grpc.Cluster cluster) {
    return LuzidCluster._(cluster);
  }

  @override
  int get hashCode => label.hashCode;

  @override
  bool operator ==(Object other) =>
      other is LuzidCluster &&
      other.runtimeType == runtimeType &&
      other.label == label;
}
